tests/e2e/lib/utils

```diff
@@ -49,6 +49,186 @@ info_message() {
+    # Arguments:                                                              #
+    #   $1 - message to output                                                #
+    ###########################################################################
+    if [ -z "${1}" ]; then
+        echo "pass_message() requires a message"
+        exit 1
+    fi
+    GREEN="\033[92m"
+    ENDCOLOR="\033[0m"
+    echo -e "[ ${GREEN}PASS${ENDCOLOR}  ] ${1}"
+}
+
+fail_message() {
+    ###########################################################################
+    # Description:                                                            #
+    # show [FAIL] in red and a message as the test failed.                    #
+    #                                                                         #
+    # Arguments:                                                              #
+    #   $1 - message to output                                                #
+    ###########################################################################
+    if [ -z "${1}" ]; then
+        echo "fail_message() requires a message"
+        exit 1
+    fi
+    RED="\033[91m"
+    ENDCOLOR="\033[0m"
+    echo -e "[ ${RED}FAIL${ENDCOLOR}  ] ${1}"
+}
+
+# Validation functions for run_test
+validate_output_pattern() {
+    local test_name="$1"
+    local validation_value="$2"
+    local stdout_file="$3"
+
+    # If no validation value provided, just check if output exists
+    if [ -z "$validation_value" ]; then
+        if [ -s "$stdout_file" ]; then
+            pass_message "$test_name - Output contains data"
+        else
+            fail_message "$test_name - Expected output data but got empty output"
+            exit 1
+        fi
+    elif grep -q "$validation_value" "$stdout_file"; then
+        pass_message "$test_name - Output contains expected pattern"
+    else
+        fail_message "$test_name - Output doesn't contain expected pattern '$validation_value'"
+        echo "Actual output (first 10 lines):"
+        head -10 < "$stdout_file"
+        exit 1
+    fi
+}
+
+validate_error_pattern() {
+    local test_name="$1"
+    local validation_value="$2"
+    local stderr_file="$3"
+
+    if [ -n "$validation_value" ] && grep -q "$validation_value" "$stderr_file"; then
+        pass_message "$test_name - Error message contains expected pattern"
+    elif [ -n "$validation_value" ]; then
+        fail_message "$test_name - Error message doesn't contain expected pattern '$validation_value'"
+        echo "Actual stderr:"
+        cat "$stderr_file"
+        exit 1
+    fi
+}
+
+validate_json_valid() {
+    local test_name="$1"
+    local stdout_file="$2"
+
+    if python3 -m json.tool < "$stdout_file" > /dev/null 2>&1; then
+        pass_message "$test_name - Valid JSON output"
+        if grep -q '{' "$stdout_file" && grep -q '}' "$stdout_file"; then
+            pass_message "$test_name - JSON contains expected braces"
+        fi
+        if [ -s "$stdout_file" ]; then
```

Comment:
just containing data it means pass? I think we need to do if grep -q ..... else fail.

Reply:
Added data to validate from with the json output